---
title: 1/27/18 - 1/29/18 Creating counts from raw reads
output: 
  html_document: 
    df_print: tibble
    toc: yes
    theme: paper
    highlight: kate
---

## 1. Extract raw data

### Location of raw data

*Raw mRNA sequencing libraries returned from the GSL are located at:

 ```linux
 /media/disk6/ARF/RNASEQ/rawreads/EWtarfiles && LGEPtarfiles
 ```

### Extracting LGEP tar files

  *Within the LGEP directory there are 18 .tar files (144 biological samples) and then an additional directory which    includes tar files from the first run of sequencing. The 18 tar files were delivered at two seperate times:
     
  *First set included 6 tar files labeled: 
      
      -Project_20150504-GSL028B-Whetten-X1POOL && X2,X3,X4,X5,X6
      
  *Second set included 12 tar files labeled: 
  
      -Project_20150615-GSL033A-Whetten-Y1POOL && 2,3,4,5,6 && Z1,Z2,Z3,Z4,Z5,Z6
    
  *The names of directores where manually edited to be:
      
      -lane01:lane06 corresponding to X1:X6 
      
      -lane07:lane12 corresponding to Y1:Y6
      
      -lane13:lane18 corresponding to Z1:Z6
     
   *Note sample in lane12 tar file 11x really belongs to lane11 so it was moved there.

### Extracting EW tar files

*Within the EW directory there are 12 .tar files (96 biological samples). The 12 tar files were delivered at two seperate times:
     
  *First set included 4 tar files labeled: 
      
      -20151203-GSL-046-B-Whetten-Pool9 && 10,11,12
     
  *Second set included 8 tar files labeled: 
  
      -GSL-047-A-Whetten-Pool1 && 2,3,4,5,6,7,8
  
  *The names of directories were manually edited to be
   
      -lane01:lane12 corresponding to Pool  
  
      
### Untar sequencing lanes

 ```linux
 cd /media/disk6/ARF/RNASEQ/rawreads/EWtarfiles
 ls *.tar > EW.tar.list
    
 while read file; do tar -xf ./$file --directory ../EWfasta/; done < EW.tar.list
 ```    

 ```linux
 cd /media/disk6/ARF/RNASEQ/rawreads/LGEPtarfiles
 ls *.tar > LGEP.tar.list
    
 while read file; do tar -xf ./$file --directory ../LGEPfasta/; done < LGEP.tar.list
 ```    

*lanes 11,12,13,14,15,16 have folders of 24 samples with 3 smaller .fasta.gz files within each folder corresponding to that sample.  These samples need to be merged for further processing to be identified as one single techincal rep.

 ```linux
 cd /media/disk6/ARF/RNASEQ/rawreads/LGEPfasta/lane11
 ls -d S* > list.lane
 while read file; do cat ./$file/*.fastq.gz > $file.fastq.gz; done < list.lane

 cd /media/disk6/ARF/RNASEQ/rawreads/LGEPfasta/lane12
 ls -d S* > list.lane
 while read file; do cat ./$file/*.fastq.gz > $file.fastq.gz; done < list.lane

 ....
 ....
 ```

## 2. Run fastqc on raw files

*Change into each of the 12 directories; create list of .fastq.gz files; make fastqc output dir and run fastqc
  
 ```linux 
 cd /media/disk6/ARF/RNASEQ/rawreads/EWfasta/lane10
 ls *.fastq.gz > lane10.list
 mkdir fastqc
 while read file; do /media/disk6/ARF/RNASEQ/software/FastQC/fastqc $file -o ./fastqc; done < lane10.list
   
 cd /media/disk6/ARF/RNASEQ/rawreads/EWfasta/lane12
 ls *.fastq.gz > lane12.list
 mkdir fastqc
 while read file; do /media/disk6/ARF/RNASEQ/software/FastQC/fastqc $file -o ./fastqc; done < lane12.list
 ...
 ...
 ```
     
 *Change into each of the 18 directories; create list of .fastq.gz files; make fastqc output dir and run fastqc
        
 ```linux 
 cd /media/disk6/ARF/RNASEQ/rawreads/LGEPfasta/lane10
 ls *.fastq.gz > lane10.list
 mkdir fastqc
 while read file; do /media/disk6/ARF/RNASEQ/software/FastQC/fastqc $file -o ./fastqc; done < lane10.list
   
   
 cd /media/disk6/ARF/RNASEQ/rawreads/EWfasta/lane12
 ls *.fastq.gz > lane12.list
 mkdir fastqc
 while read file; do /media/disk6/ARF/RNASEQ/software/FastQC/fastqc $file -o ./fastqc; done < lane12.list
 ...
 ...
 ```
      
**Noticed there is rRNA contamination in Sample 71 in EW lane 09.***


## 3. Remove rRNA contamination

*Accessed www.arb-silva.de and searched for Pinus taeda

*805 Accessions showed up in results and were downloaded.  The original .tgz is located at 

 ```linux
 /media/disk6/ARF/RNASEQ/software/arb-silva.de_2018-01-28_id495190.tgz
 ```

*The .fasta file inside the .tgz was placed in:

 ```linux
 /media/disk6/ARF/RNASEQ/software/bbmap/resources
 ```

*bbDuk can be used to do overall quality trim, as well as trim first 10 bases, remove adapters, and matches to rRNA.

*Test it out on EW sample 84 from lane10

```linux
./software/bbmap/bbduk.sh -Xmx2g in=./rawreads/EWfasta/lane10/84_S29_L002_R1_001.fastq.gz out=./8429l2.fq.gz      ref=./software/bbmap/resources/truseq.fa.gz,./software/bbmap/resources/rRNA_lob_pine.fasta forcetrimleft=10 t=1 ktrim=r k=25 mink=12 hdist=1 qtrim=rl trimq=20 minlength=50 maxlength=125
```
*Initially Sample 84 had rRNA and adapter contamination.  After running bbduk there was still one overresprestned sequence which when blasted hit to a tumor response protein.  Since this is biologicaly relevenat it should be OK to stay.  We can use bbduk now to process all raw reads.

*Article suggest low to minimium quality trimming works best: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4766705/
*Also minimum read length for DE expression roughly 50 works: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4531809/

*First run on the EW technical reps a single sample at a time

```linux
cd /media/disk6/ARF/RNASEQ
while read file; do ./software/bbmap/bbduk.sh -Xmx2g in=./rawreads/EWfasta/lane02/$file     out=./trimmedfiltreads/EW/lane02/$file ref=./software/bbmap/resources/truseq.fa.gz,./software/bbmap/resources/rRNA_lob_pine.fasta forcetrimleft=10 t=5 ktrim=r k=25 mink=12 hdist=1 hdist2=0 qtrim=rl trimq=20 minlength=50 ; done < ./rawreads/EWfasta/lane02/lane02.list

while read file; do ./software/bbmap/bbduk.sh -Xmx3g in=./rawreads/EWfasta/lane02/$file out=./trimmedfiltreads/EW/lane02/$file ref=./software/bbmap/resources/truseq.fa.gz,./software/bbmap/resources/rRNA_lob_pine.fasta forcetrimleft=10 t=5 ktrim=r k=25 mink=12 hdist=1 hdist2=0 qtrim=rl trimq=20 minlength=50 ; done < ./rawreads/EWfasta/lane02/lane02.list
```

*Output is located at:
```linux
/media/disk6/ARF/RNASEQ/trimmedfiltreads/EW/lane01 
```

*Now run on the LGEP technical reps a single sample at a time

```linux
while read file; do ./software/bbmap/bbduk.sh -Xmx3g in=./rawreads/LGEPfasta/lane01/$file out=./trimmedfiltreads/LGEP/lane01/$file ref=./software/bbmap/resources/truseq.fa.gz,./software/bbmap/resources/rRNA_lob_pine.fasta forcetrimleft=10 t=5 ktrim=r k=25 mink=12 hdist=1 hdist2=0 qtrim=rl trimq=20 minlength=50 ; done < ./rawreads/LGEPfasta/lane01/lane01.list

...

while read file; do ./software/bbmap/bbduk.sh -Xmx3g in=./rawreads/LGEPfasta/lane09/$file out=./trimmedfiltreads/LGEP/lane09/$file ref=./software/bbmap/resources/truseq.fa.gz,./software/bbmap/resources/rRNA_lob_pine.fasta forcetrimleft=10 t=5 ktrim=r k=25 mink=12 hdist=1 hdist2=0 qtrim=rl trimq=20 minlength=50 ; done < ./rawreads/LGEPfasta/lane09/lane09.list
```

*Multiple lanes ran one after the other:

```linux
while read file; do ./software/bbmap/bbduk.sh -Xmx3g in=./rawreads/LGEPfasta/lane07/$file out=./trimmedfiltreads/LGEP/lane07/$file ref=./software/bbmap/resources/truseq.fa.gz,./software/bbmap/resources/rRNA_lob_pine.fasta forcetrimleft=10 t=5 ktrim=r k=25 mink=12 hdist=1 hdist2=0 qtrim=rl trimq=20 minlength=50 ; done < ./rawreads/LGEPfasta/lane07/lane07.list; while read file; do ./software/bbmap/bbduk.sh -Xmx3g in=./rawreads/LGEPfasta/lane08/$file out=./trimmedfiltreads/LGEP/lane08/$file ref=./software/bbmap/resources/truseq.fa.gz,./software/bbmap/resources/rRNA_lob_pine.fasta forcetrimleft=10 t=5 ktrim=r k=25 mink=12 hdist=1 hdist2=0 qtrim=rl trimq=20 minlength=50 ; done < ./rawreads/LGEPfasta/lane08/lane08.list
```

All lanes are processed the same way by changing the lane numbers in the bbduk command.

Note:After processing Sample_101b in lane10 for LGEP file was much smaller, re-ran again on that file at a later time so file date is newer.

## 4. FastQC can be re-done on filtered and trimmed files

## 5. Aligning reads to 86K transcriptome with Salmon

*First index the 86k transcriptome:

*p = 5 (five threads)

*type = quasi (quasi-based mapping recommended)

*k = 25 (A k of 31 is good for minimum read length = 75, but our minimum read length is 50)

*perfectHash (takes a little longer to build index but faster for quant command)

```linux
./software/Salmon/bin/salmon index -t ./references/Pita.86ktxptome.fasta -i ./references/Pita.86k -p 5 --perfectHash --type "quasi" -k 25
```

*Salmon fld flags that it says are important for single end reads.  Need to loook into this further.

*Appears that this is a requirement of all kmer based methods and probably corresponds to the cDNA input size, therefore the default setting should be OK.

*p = 40 (use 40 threads)

*l = U (library is unstranded)

*numBootstraps = 2

```linux
Salmon has the ability to optionally compute bootstrapped abundance estimates. This is done by resampling (with replacement) from the counts assigned to the fragment equivalence classes, and then re-running the optimization procedure, either the EM or VBEM, for each such sample. The values of these different bootstraps allows us to assess technical variance in the main abundance estimates we produce. Such estimates can be useful for downstream (e.g. differential expression) tools that can make use of such uncertainty estimates. This option takes a positive integer that dictates the number of bootstrap samples to compute. The more samples computed, the better the estimates of varaiance, but the more computation (and time) required.
```

*seqBias

```linux
Passing the --seqBias flag to Salmon will enable it to learn and correct for sequence-specific biases in the input data. Specifically, this model will attempt to correct for random hexamer priming bias, which results in the preferential sequencing of fragments starting with certain nucleotide motifs. By default, Salmon learns the sequence-specific bias parameters using 1,000,000 reads from the beginning of the input. If you wish to change the number of samples from which the model is learned, you can use the --numBiasSamples parameter. Salmon uses a variable-length Markov Model (VLMM) to model the sequence specific biases at both the 5’ and 3’ end of sequenced fragments. This methodology generally follows that of Roberts et al. [2], though some details of the VLMM differ.
```


``` linux
cd /media/disk6/ARF/RNASEQ

while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/EW/lane01/$file -i ./references/Pita.86k -o ./counts/86kSalmon/EW/lane01/$file -l U --numBootstraps 2; done < ./rawreads/EWfasta/lane01/lane01.list

while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane09/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane09/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane09/lane09.list; while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane10/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane10/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane10/lane10.list; while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane11/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane11/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane11/lane11.list; while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane12/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane12/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane12/lane12.list; while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane13/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane13/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane13/lane13.list;

while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane14/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane14/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane14/lane14.list; while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane15/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane15/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane15/lane15.list; while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane16/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane16/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane16/lane16.list; while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane17/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane17/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane17/lane17.list; while read file; do ./software/Salmon/bin/salmon quant --seqBias -r ./trimmedfiltreads/LGEP/lane18/$file -i ./references/Pita.86k -o ./counts/86kSalmon/LGEP/lane18/$file -l U -p 40 --numBootstraps 2; done < ./rawreads/LGEPfasta/lane18/lane18.list;
``` 
*Then align counts and place in counts folder
  
## 7. Aligning reads to 83K transcriptome with Salmon  
```linux
  ./software/Salmon/bin/salmon index -t ./references/Pita.83k.fa -i ./references/Pita.83k -p 5 --perfectHash --type "quasi" -k 25
```  
 


  
